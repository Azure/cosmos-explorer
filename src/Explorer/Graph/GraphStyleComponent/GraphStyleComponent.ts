import * as Constants from "../../../Common/Constants";
import * as ViewModels from "../../../Contracts/ViewModels";
import { WaitsForTemplateViewModel } from "../../WaitsForTemplateViewModel";

/**
 * Parameters for this component
 */
export interface GraphStyleParams {
  config: ViewModels.GraphConfigUiData;
  firstFieldHasFocus?: ko.Observable<boolean>;

  /**
   * Callback triggered when the template is bound to the component (for testing purposes)
   */
  onTemplateReady?: () => void;
}

class GraphStyleViewModel extends WaitsForTemplateViewModel {
  private params: GraphStyleParams;

  public constructor(params: GraphStyleParams) {
    super();
    super.onTemplateReady((isTemplateReady: boolean) => {
      if (isTemplateReady && params.onTemplateReady) {
        params.onTemplateReady();
      }
    });

    this.params = params;
  }

  public onAllNeighborsKeyPress = (source: any, event: KeyboardEvent): boolean => {
    if (event.keyCode === Constants.KeyCodes.Space || event.keyCode === Constants.KeyCodes.Enter) {
      this.params.config.showNeighborType(ViewModels.NeighborType.BOTH);
      event.stopPropagation();
      return false;
    }

    return true;
  };

  public onSourcesKeyPress = (source: any, event: KeyboardEvent): boolean => {
    if (event.keyCode === Constants.KeyCodes.Space || event.keyCode === Constants.KeyCodes.Enter) {
      this.params.config.showNeighborType(ViewModels.NeighborType.SOURCES_ONLY);
      event.stopPropagation();
      return false;
    }

    return true;
  };

  public onTargetsKeyPress = (source: any, event: KeyboardEvent): boolean => {
    if (event.keyCode === Constants.KeyCodes.Space || event.keyCode === Constants.KeyCodes.Enter) {
      this.params.config.showNeighborType(ViewModels.NeighborType.TARGETS_ONLY);
      event.stopPropagation();
      return false;
    }

    return true;
  };
}

const template = `
<div id="graphStyle" class="graphStyle" data-bind="setTemplateReady: true, with:params.config">
    <div class="seconddivpadding">
        <p>Show vertex (node) as</p>
        <select id="nodeCaptionChoices" class="formTree paneselect" required data-bind="options:nodeProperties,
                                            value:nodeCaptionChoice, hasFocus: $parent.params.firstFieldHasFocus"></select>
    </div>
    <div class="seconddivpadding">
        <p>Map this property to node color</p>
        <select id="nodeColorKeyChoices" class="formTree paneselect" required data-bind="options:nodePropertiesWithNone,
                                                                            value:nodeColorKeyChoice"></select>
    </div>
    <div class="seconddivpadding">
        <p>Map this property to node icon</p>
        <select id="nodeIconChoices" class="formTree paneselect" required data-bind="options:nodePropertiesWithNone,
                                                                            value:nodeIconChoice"></select>
        <input type="text" data-bind="value:nodeIconSet" placeholder="Icon set: blank for collection id" class="nodeIconSet" autocomplete="off" />
    </div>

    <p class="seconddivpadding">Show</p>

    <div class="tabs">
        <div class="tab">
            <input type="radio" id="tab11" name="graphneighbortype" class="radio" data-bind="checkedValue:2, checked:showNeighborType" />
            <label for="tab11" tabindex="0" data-bind="event: { keypress: $parent.onAllNeighborsKeyPress }">All neighbors</label>
        </div>
        <div class="tab">
            <input type="radio" id="tab12" name="graphneighbortype" class="radio" data-bind="checkedValue:0, checked:showNeighborType" />
            <label for="tab12" tabindex="0" data-bind="event: { keypress: $parent.onSourcesKeyPress }">Sources</label>
        </div>
        <div class="tab">
            <input type="radio" id="tab13" name="graphneighbortype" class="radio" data-bind="checkedValue:1, checked:showNeighborType" />
            <label for="tab13" tabindex="0" data-bind="event: { keypress: $parent.onTargetsKeyPress }">Targets</label>
        </div>
    </div>
</div>`;

export const GraphStyleComponent = {
  viewModel: GraphStyleViewModel,
  template
};
