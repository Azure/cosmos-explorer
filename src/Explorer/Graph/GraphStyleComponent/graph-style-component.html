<div id="graphStyle" class="graphStyle" data-bind="setTemplateReady: true, with:params.config">
  <div class="seconddivpadding">
    <p>Show vertex (node) as</p>
    <select
      id="nodeCaptionChoices"
      class="formTree paneselect"
      required
      data-bind="options:nodeProperties,
                                            value:nodeCaptionChoice, hasFocus: $parent.params.firstFieldHasFocus"
    ></select>
  </div>
  <div class="seconddivpadding">
    <p>Map this property to node color</p>
    <select
      id="nodeColorKeyChoices"
      class="formTree paneselect"
      required
      data-bind="options:nodePropertiesWithNone,
                                                                            value:nodeColorKeyChoice"
    ></select>
  </div>
  <div class="seconddivpadding">
    <p>Map this property to node icon</p>
    <select
      id="nodeIconChoices"
      class="formTree paneselect"
      required
      data-bind="options:nodePropertiesWithNone,
                                                                            value:nodeIconChoice"
    ></select>
    <input
      type="text"
      data-bind="value:nodeIconSet"
      placeholder="Icon set: blank for collection id"
      class="nodeIconSet"
      autocomplete="off"
    />
  </div>

  <p class="seconddivpadding">Show</p>

  <div class="tabs">
    <div class="tab">
      <input
        type="radio"
        id="tab11"
        name="graphneighbortype"
        class="radio"
        data-bind="checkedValue:2, checked:showNeighborType"
      />
      <label for="tab11">All neighbors</label>
    </div>
    <div class="tab">
      <input
        type="radio"
        id="tab12"
        name="graphneighbortype"
        class="radio"
        data-bind="checkedValue:0, checked:showNeighborType"
      />
      <label for="tab12">Sources</label>
    </div>
    <div class="tab">
      <input
        type="radio"
        id="tab13"
        name="graphneighbortype"
        class="radio"
        data-bind="checkedValue:1, checked:showNeighborType"
      />
      <label for="tab13">Targets</label>
    </div>
  </div>
</div>
