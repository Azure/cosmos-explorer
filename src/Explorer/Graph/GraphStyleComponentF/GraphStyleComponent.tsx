import { ChoiceGroup, Dropdown, IChoiceGroupOption, IDropdownStyles, Stack, TextField } from "office-ui-fabric-react";
import React, { FunctionComponent } from "react";
import { IGraphConfigUiData, NeighborType } from "../../../Contracts/ViewModels";

interface GraphStyleProps {
  config: IGraphConfigUiData;
  // firstFieldHasFocus: boolean;
}

export const GraphStyleComponentF: FunctionComponent<GraphStyleProps> = ({
  config
  // firstFieldHasFocus,
}: GraphStyleProps): JSX.Element => {
  // const [firstFieldHasFocusState, setFirstFieldHasFocus] = useState(firstFieldHasFocus);

  const nodePropertiesOptions = config.nodeProperties.map((nodeProperty) => ({
    key: nodeProperty,
    text: nodeProperty,
  }));

  // const nodePropertiesOptions = [
  //   { key: "1", text: "Test1" },
  //   { key: "2", text: "Test2" },
  //   { key: "3", text: "Test3" },
  // ];

  const nodePropertiesWithNoneOptions = config.nodePropertiesWithNone.map((nodePropertyWithNone) => ({
    key: nodePropertyWithNone,
    text: nodePropertyWithNone,
  }));

  // const nodePropertiesWithNoneOptions = [
  //   { key: "1", text: "Node1" },
  //   { key: "2", text: "Node2" },
  //   { key: "3", text: "Node2" },
  // ];

  const showNeighborTypeOptions: IChoiceGroupOption[] = [
    { key: NeighborType.BOTH.toString(), text: "All neighbors" },
    { key: NeighborType.SOURCES_ONLY.toString(), text: "Sources" },
    { key: NeighborType.TARGETS_ONLY.toString(), text: "Targets" },
  ];

  const dropdownStyles: Partial<IDropdownStyles> = { dropdown: { height: 32, marginRight: 10 } };
  const choiceButtonStyles = {
    flexContainer: [
      {
        selectors: {
          ".ms-ChoiceField-wrapper label": {
            fontSize: 14,
            paddingTop: 0,
          },
        },
      },
    ],
  };

  return (
    <Stack>
      <div id="graphStyle" className="graphStyle">
        <div className="seconddivpadding">
          <Dropdown label="Show vertex (node) as" options={nodePropertiesOptions} required styles={dropdownStyles} />
        </div>
        <div className="seconddivpadding">
          <Dropdown
            label="Map this property to node color"
            options={nodePropertiesWithNoneOptions}
            required
            styles={dropdownStyles}
          />
        </div>
        <div className="seconddivpadding">
          <Stack horizontal>
            <Dropdown
              label="Map this property to node icon"
              options={nodePropertiesWithNoneOptions}
              required
              styles={dropdownStyles}
            />
            <Stack.Item align="end">
              <TextField
                // value={nodeIconSet}
                // label=" "
                placeholder="Icon set: blank for collection id"
                inputClassName="nodeIconSet"
                autoComplete="off"
              />
            </Stack.Item>
          </Stack>
        </div>
        <p className="seconddivpadding">Show</p>

        <ChoiceGroup
          styles={choiceButtonStyles}
          options={showNeighborTypeOptions}
          onChange={() => config.setValues()}
        />
      </div>
    </Stack>
  );
};
