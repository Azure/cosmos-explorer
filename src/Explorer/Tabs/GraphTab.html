<div class="graphExplorerContainer" role="tabpanel" data-bind="react:graphExplorerAdapter, attr:{ id: tabId }"></div>
