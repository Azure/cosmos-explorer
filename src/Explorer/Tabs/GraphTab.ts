import * as ko from "knockout";
import * as Q from "q";
import * as ViewModels from "../../Contracts/ViewModels";
import TabsBase from "./TabsBase";
import { GraphExplorerAdapter } from "../Graph/GraphExplorerComponent/GraphExplorerAdapter";
import { GraphAccessor, GraphExplorerError } from "../Graph/GraphExplorerComponent/GraphExplorer";
import NewVertexIcon from "../../../images/NewVertex.svg";
import StyleIcon from "../../../images/Style.svg";
import GraphStylingPane from "../Panes/GraphStylingPane";
import NewVertexPane from "../Panes/NewVertexPane";
import { DatabaseAccount } from "../../Contracts/DataModels";
import { CommandButtonComponentProps } from "../Controls/CommandButton/CommandButtonComponent";

export interface GraphIconMap {
  [key: string]: { data: string; format: string };
}

export interface GraphConfig {
  nodeColor: ko.Observable<string>;
  nodeColorKey: ko.Observable<string>; // map property to node color. Takes precedence over nodeColor unless null
  linkColor: ko.Observable<string>;
  showNeighborType: ko.Observable<ViewModels.NeighborType>;
  nodeCaption: ko.Observable<string>;
  nodeSize: ko.Observable<number>;
  linkWidth: ko.Observable<number>;
  nodeIconKey: ko.Observable<string>;
  iconsMap: ko.Observable<GraphIconMap>;
}

interface GraphTabOptions extends ViewModels.TabOptions {
  account: DatabaseAccount;
  masterKey: string;
  collectionId: string;
  databaseId: string;
  collectionPartitionKeyProperty: string;
}

export default class GraphTab extends TabsBase {
  // Graph default configuration
  public static readonly DEFAULT_NODE_CAPTION = "id";
  private static readonly LINK_COLOR = "#aaa";
  private static readonly NODE_SIZE = 10;
  private static readonly NODE_COLOR = "orange";
  private static readonly LINK_WIDTH = 1;
  private graphExplorerAdapter: GraphExplorerAdapter;
  private isNewVertexDisabled: ko.Observable<boolean>;
  private isPropertyEditing: ko.Observable<boolean>;
  private isGraphDisplayed: ko.Observable<boolean>;
  private graphAccessor: GraphAccessor;
  private graphConfig: GraphConfig;
  private graphConfigUiData: ViewModels.GraphConfigUiData;
  private isFilterQueryLoading: ko.Observable<boolean>;
  private isValidQuery: ko.Observable<boolean>;
  private newVertexPane: NewVertexPane;
  private graphStylingPane: GraphStylingPane;
  private collectionPartitionKeyProperty: string;

  constructor(options: GraphTabOptions) {
    super(options);

    this.newVertexPane = options.collection && options.collection.container.newVertexPane;
    this.graphStylingPane = options.collection && options.collection.container.graphStylingPane;
    this.collectionPartitionKeyProperty = options.collectionPartitionKeyProperty;

    this.isNewVertexDisabled = ko.observable(false);
    this.isPropertyEditing = ko.observable(false);
    this.isGraphDisplayed = ko.observable(false);
    this.graphAccessor = null;
    this.graphConfig = GraphTab.createGraphConfig();
    // TODO Merge this with this.graphConfig
    this.graphConfigUiData = GraphTab.createGraphConfigUiData(this.graphConfig);
    this.graphExplorerAdapter = new GraphExplorerAdapter({
      onGraphAccessorCreated: (instance: GraphAccessor): void => {
        this.graphAccessor = instance;
      },
      onIsNewVertexDisabledChange: (isDisabled: boolean): void => {
        this.isNewVertexDisabled(isDisabled);
        this.updateNavbarWithTabsButtons();
      },
      onIsPropertyEditing: (isEditing: boolean) => {
        this.isPropertyEditing(isEditing);
        this.updateNavbarWithTabsButtons();
      },
      onIsGraphDisplayed: (isDisplayed: boolean) => {
        this.isGraphDisplayed(isDisplayed);
        this.updateNavbarWithTabsButtons();
      },
      onResetDefaultGraphConfigValues: () => this.setDefaultGraphConfigValues(),
      graphConfig: this.graphConfig,
      graphConfigUiData: this.graphConfigUiData,
      onIsFilterQueryLoading: (isFilterQueryLoading: boolean): void => this.isFilterQueryLoading(isFilterQueryLoading),
      onIsValidQuery: (isValidQuery: boolean): void => this.isValidQuery(isValidQuery),
      collectionPartitionKeyProperty: options.collectionPartitionKeyProperty,
      graphBackendEndpoint: GraphTab.getGremlinEndpoint(options.account),
      databaseId: options.databaseId,
      collectionId: options.collectionId,
      masterKey: options.masterKey,
      onLoadStartKey: options.onLoadStartKey,
      onLoadStartKeyChange: (onLoadStartKey: number): void => {
        if (onLoadStartKey == null) {
          this.onLoadStartKey = null;
        }
      },
      resourceId: options.account.id
    });

    this.isFilterQueryLoading = ko.observable(false);
    this.isValidQuery = ko.observable(true);
  }

  public static getGremlinEndpoint(account: DatabaseAccount): string {
    return account.properties.gremlinEndpoint
      ? GraphTab.sanitizeHost(account.properties.gremlinEndpoint)
      : `${account.name}.graphs.azure.com:443/`;
  }

  public onTabClick(): Q.Promise<any> {
    return super.onTabClick().then(() => {
      this.collection.selectedSubnodeKind(ViewModels.CollectionTabKind.Graph);
    });
  }

  /**
   * Removing leading http|https and remove trailing /
   * @param url
   * @return
   */
  private static sanitizeHost(url: string): string {
    if (!url) {
      return url;
    }
    return url.replace(/^(http|https):\/\//, "").replace(/\/$/, "");
  }

  /* Command bar */
  private showNewVertexEditor(): void {
    this.newVertexPane.open();
    this.newVertexPane.setPartitionKeyProperty(this.collectionPartitionKeyProperty);
    // TODO Must update GraphExplorer properties
    this.newVertexPane.subscribeOnSubmitCreate((result: ViewModels.NewVertexData) => {
      this.newVertexPane.formErrors(null);
      this.newVertexPane.formErrorsDetails(null);
      this.graphAccessor.addVertex(result).then(
        () => {
          this.newVertexPane.cancel();
        },
        (error: GraphExplorerError) => {
          this.newVertexPane.formErrors(error.title);
          if (!!error.details) {
            this.newVertexPane.formErrorsDetails(error.details);
          }
        }
      );
    });
  }
  public openStyling() {
    this.setDefaultGraphConfigValues();
    // Update the styling pane with this instance
    this.graphStylingPane.setData(this.graphConfigUiData);
    this.graphStylingPane.open();
  }

  public static createGraphConfig(): GraphConfig {
    return {
      nodeColor: ko.observable(GraphTab.NODE_COLOR),
      nodeColorKey: ko.observable(null),
      linkColor: ko.observable(GraphTab.LINK_COLOR),
      showNeighborType: ko.observable(ViewModels.NeighborType.TARGETS_ONLY),
      nodeCaption: ko.observable(GraphTab.DEFAULT_NODE_CAPTION),
      nodeSize: ko.observable(GraphTab.NODE_SIZE),
      linkWidth: ko.observable(GraphTab.LINK_WIDTH),
      nodeIconKey: ko.observable(null),
      iconsMap: ko.observable({})
    };
  }

  public static createGraphConfigUiData(graphConfig: GraphConfig): ViewModels.GraphConfigUiData {
    return {
      showNeighborType: ko.observable(graphConfig.showNeighborType()),
      nodeProperties: ko.observableArray([]),
      nodePropertiesWithNone: ko.observableArray([]),
      nodeCaptionChoice: ko.observable(graphConfig.nodeCaption()),
      nodeColorKeyChoice: ko.observable(graphConfig.nodeColorKey()),
      nodeIconChoice: ko.observable(graphConfig.nodeIconKey()),
      nodeIconSet: ko.observable(null)
    };
  }

  /**
   * Make sure graph config values are not null
   */
  private setDefaultGraphConfigValues() {
    // Assign default values if null
    if (this.graphConfigUiData.nodeCaptionChoice() === null && this.graphConfigUiData.nodeProperties().length > 1) {
      this.graphConfigUiData.nodeCaptionChoice(this.graphConfigUiData.nodeProperties()[0]);
    }
    if (
      this.graphConfigUiData.nodeColorKeyChoice() === null &&
      this.graphConfigUiData.nodePropertiesWithNone().length > 1
    ) {
      this.graphConfigUiData.nodeColorKeyChoice(this.graphConfigUiData.nodePropertiesWithNone()[0]);
    }
    if (
      this.graphConfigUiData.nodeIconChoice() === null &&
      this.graphConfigUiData.nodePropertiesWithNone().length > 1
    ) {
      this.graphConfigUiData.nodeIconChoice(this.graphConfigUiData.nodePropertiesWithNone()[0]);
    }
  }
  protected getTabsButtons(): CommandButtonComponentProps[] {
    const label = "New Vertex";
    const buttons: CommandButtonComponentProps[] = [
      {
        iconSrc: NewVertexIcon,
        iconAlt: label,
        onCommandClick: () => this.showNewVertexEditor(),
        commandButtonLabel: label,
        ariaLabel: label,
        hasPopup: false,
        disabled: this.isNewVertexDisabled()
      }
    ];
    buttons.push();
    if (this.isGraphDisplayed()) {
      const label = "Style";
      buttons.push({
        iconSrc: StyleIcon,
        iconAlt: label,
        onCommandClick: () => this.openStyling(),
        commandButtonLabel: label,
        ariaLabel: label,
        hasPopup: false,
        disabled: this.isPropertyEditing()
      });
    }
    return buttons;
  }
  protected buildCommandBarOptions(): void {
    ko.computed(() => ko.toJSON([this.isNewVertexDisabled])).subscribe(() => this.updateNavbarWithTabsButtons());
    this.updateNavbarWithTabsButtons();
  }
}
