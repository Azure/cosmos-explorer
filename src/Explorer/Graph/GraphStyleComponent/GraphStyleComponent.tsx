import { ChoiceGroup, Dropdown, IChoiceGroupOption, IDropdownOption, IDropdownStyles, Stack } from "@fluentui/react";
import React, { FunctionComponent, useEffect, useState } from "react";
import { IGraphConfigUiData, NeighborType } from "../../../Contracts/ViewModels";
import { IGraphConfig } from "../../Tabs/GraphTab";
const IGraphConfigType = {
  NODE_CAPTION: "NODE_CAPTION",
  NODE_COLOR: "NODE_COLOR",
  NODE_ICON: "NODE_ICON",
  SHOW_NEIGHBOR_TYPE: "SHOW_NEIGHBOR_TYPE",
};
export interface GraphStyleProps {
  igraphConfig: IGraphConfig;
  igraphConfigUiData: IGraphConfigUiData;
  getValues: (igraphConfig?: IGraphConfig) => void;
}

export const GraphStyleComponent: FunctionComponent<GraphStyleProps> = ({
  igraphConfig,
  igraphConfigUiData,
  getValues,
}: GraphStyleProps): JSX.Element => {
  const [igraphConfigState, setIGraphConfig] = useState<IGraphConfig>(igraphConfig);
  const [selected, setSelected] = useState<boolean>(false);

  const nodePropertiesOptions = igraphConfigUiData.nodeProperties.map((nodeProperty) => ({
    key: nodeProperty,
    text: nodeProperty,
  }));

  const nodePropertiesWithNoneOptions = igraphConfigUiData.nodePropertiesWithNone.map((nodePropertyWithNone) => ({
    key: nodePropertyWithNone,
    text: nodePropertyWithNone,
  }));

  const showNeighborTypeOptions: IChoiceGroupOption[] = [
    { key: NeighborType.BOTH.toString(), text: "All neighbors" },
    { key: NeighborType.SOURCES_ONLY.toString(), text: "Sources" },
    { key: NeighborType.TARGETS_ONLY.toString(), text: "Targets" },
  ];

  const dropdownStyles: Partial<IDropdownStyles> = {
    dropdown: { height: 32, marginRight: 10 },
  };
  const choiceButtonStyles = {
    flexContainer: [
      {
        selectors: {
          ".ms-ChoiceField-wrapper label": {
            fontSize: 14,
            paddingTop: 0,
          },
        },
      },
    ],
  };

  useEffect(() => {
    if (selected) {
      getValues(igraphConfigState);
    }
    //eslint-disable-next-line
  }, [igraphConfigState]);

  const handleOnChange = (val: string, igraphConfigType: string) => {
    switch (igraphConfigType) {
      case IGraphConfigType.NODE_CAPTION:
        setSelected(true);
        setIGraphConfig({
          ...igraphConfigState,
          nodeCaption: val,
        });
        break;
      case IGraphConfigType.NODE_COLOR:
        setSelected(true);
        setIGraphConfig({
          ...igraphConfigState,
          nodeColorKey: val,
        });
        break;
      case IGraphConfigType.SHOW_NEIGHBOR_TYPE:
        setSelected(true);
        setIGraphConfig({
          ...igraphConfigState,
          showNeighborType: parseInt(val),
        });
        break;
    }
  };
  return (
    <Stack>
      <div id="graphStyle" className="graphStyle">
        <div className="seconddivpadding">
          <Dropdown
            label="Show vertex (node) as"
            options={nodePropertiesOptions}
            required
            selectedKey={igraphConfigState.nodeCaption}
            styles={dropdownStyles}
            onChange={(_, options: IDropdownOption) =>
              handleOnChange(options.key.toString(), IGraphConfigType.NODE_CAPTION)
            }
          />
        </div>
        <div className="seconddivpadding">
          <Dropdown
            label="Map this property to node color"
            options={nodePropertiesWithNoneOptions}
            required
            selectedKey={igraphConfigState.nodeColorKey}
            styles={dropdownStyles}
            onChange={(_, options: IDropdownOption) =>
              handleOnChange(options.key.toString(), IGraphConfigType.NODE_COLOR)
            }
          />
        </div>

        <div className="seconddivpadding">
          <ChoiceGroup
            label="Show"
            styles={choiceButtonStyles}
            options={showNeighborTypeOptions}
            selectedKey={igraphConfigState.showNeighborType.toString()}
            onChange={(_, options: IChoiceGroupOption) =>
              handleOnChange(options.key.toString(), IGraphConfigType.SHOW_NEIGHBOR_TYPE)
            }
          />
        </div>
      </div>
    </Stack>
  );
};
