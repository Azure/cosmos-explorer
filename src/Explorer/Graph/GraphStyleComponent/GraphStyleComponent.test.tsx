import { render, screen } from "@testing-library/react";
import React from "react";
import * as ViewModels from "../../../Contracts/ViewModels";
import { IGraphConfig } from "../../Tabs/GraphTab";
import { GraphStyleComponent, GraphStyleProps } from "./GraphStyleComponent";

describe("Graph Style Component", () => {
  let fakeGraphConfig: IGraphConfig;
  let fakeGraphConfigUiData: ViewModels.IGraphConfigUiData;
  let props: GraphStyleProps;
  beforeEach(() => {
    fakeGraphConfig = {
      nodeColor: "orange",
      nodeColorKey: "node2",
      linkColor: "orange",
      showNeighborType: 0,
      nodeCaption: "node1",
      nodeSize: 10,
      linkWidth: 1,
      nodeIconKey: undefined,
      iconsMap: {},
    };
    fakeGraphConfigUiData = {
      nodeCaptionChoice: "node1",
      nodeIconChoice: undefined,
      nodeColorKeyChoice: "node2",
      nodeIconSet: undefined,
      nodeProperties: ["node1", "node2", "node3"],
      nodePropertiesWithNone: ["none", "node1", "node2", "node3"],
      showNeighborType: undefined,
    };
    props = {
      igraphConfig: fakeGraphConfig,
      igraphConfigUiData: fakeGraphConfigUiData,
      getValues: (): void => undefined,
    };

    render(<GraphStyleComponent {...props} />);
  });

  it("should render default property", () => {
    const { asFragment } = render(<GraphStyleComponent {...props} />);
    expect(asFragment).toMatchSnapshot();
  });

  it("should render node properties dropdown list ", () => {
    const dropDownList = screen.getByText("Show vertex (node) as");
    expect(dropDownList).toBeDefined();
  });

  it("should render Map this property to node color dropdown list", () => {
    const nodeColorDropdownList = screen.getByText("Map this property to node color");
    expect(nodeColorDropdownList).toBeDefined();
  });

  it("should render show neighbor options", () => {
    const nodeShowNeighborOptions = screen.getByText("Show");
    expect(nodeShowNeighborOptions).toBeDefined();
  });

  it("should call handleOnChange method", () => {
    const handleOnChange = jest.fn();
    const nodeCaptionDropdownList = screen.getByText("Show vertex (node) as");
    nodeCaptionDropdownList.onchange = handleOnChange();
    expect(handleOnChange).toHaveBeenCalled();
  });
});
