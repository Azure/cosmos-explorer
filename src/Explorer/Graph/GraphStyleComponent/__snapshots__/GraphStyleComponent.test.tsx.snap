// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Style Component should render default property 1`] = `[Function]`;
